src/Renderer/LayeredMaterial.ts

```diff
@@ -0,0 +1,542 @@
+    minBorderDistance: number;
+
+    // Debug
+    showOutline: boolean,
+    outlineWidth: number,
+    outlineColors: THREE.Color[]
+
+    // Elevation layers
+    elevationLayers: Array<StructElevationLayer>,
+    elevationTextures: Array<THREE.Texture>,
+    elevationOffsetScales: Array<THREE.Vector4>,
+    elevationTextureCount: number,
+
+    // Color layers
+    colorLayers: Array<StructColorLayer>,
+    colorTextures: Array<THREE.Texture>,
+    colorOffsetScales: Array<THREE.Vector4>,
+    colorTextureCount: number,
+}
+
+let nbSamplers: [number, number] | undefined;
+const fragmentShader: string[] = [];
+
+/** Replacing the default uniforms dynamic type with our own static map. */
+export type LayeredMaterialParameters =
+    Omit<THREE.ShaderMaterialParameters, 'uniforms'>
+    & { uniforms?: MappedUniforms<LayeredMaterialRawUniforms> };
+
+type DefineMapping<Prefix extends string, Mapping extends Record<string, unknown>> = {
+    [Name in Extract<keyof Mapping, string> as `${Prefix}_${Name}`]: Mapping[Name]
+};
+
+/** Fills in a Partial object's field and narrows the type accordingly. */
+function fillInProp<
+    Obj extends Partial<Record<PropertyKey, unknown>>,
+    Name extends keyof Obj,
+    Value extends Obj[Name],
+>(
+    obj: Obj,
+    name: Name,
+    value: Value,
+): asserts obj is Obj & { [P in Name]: Value } {
+    if (obj[name] === undefined) {
+        (obj as Record<Name, Value>)[name] = value;
+    }
+}
+
+type ElevationModeDefines = DefineMapping<'ELEVATION', typeof ELEVATION_MODES>;
+type RenderModeDefines = DefineMapping<'MODE', typeof RenderMode.MODES>;
+type LayeredMaterialDefines = {
+    NUM_VS_TEXTURES: number;
+    NUM_FS_TEXTURES: number;
+    USE_FOG: number;
+    NUM_CRS: number;
+    DEBUG: number;
+    MODE: number;
+} & ElevationModeDefines
+    & RenderModeDefines;
+
+/**
+ * Initialiszes elevation and render mode defines and narrows the type
+ * accordingly.
+ */
+function initModeDefines(
+    defines: Partial<LayeredMaterialDefines>,
+): asserts defines is Partial<LayeredMaterialDefines> & ElevationModeDefines & RenderModeDefines {
+    (Object.keys(ELEVATION_MODES) as (keyof typeof ELEVATION_MODES)[])
+        .forEach(key => fillInProp(defines, `ELEVATION_${key}`, ELEVATION_MODES[key]));
+    (Object.keys(RenderMode.MODES) as (keyof typeof RenderMode.MODES)[])
+        .forEach(key => fillInProp(defines, `MODE_${key}`, RenderMode.MODES[key]));
+}
+
+/** Material that handles the overlap of multiple raster tiles. */
+export class LayeredMaterial extends THREE.ShaderMaterial {
+    private _visible = true;
+
+    public colorTiles: RasterColorTile[];
+    public elevationTile: RasterElevationTile | undefined;
+
+    public colorTileIds: string[];
+    public elevationTileId: string | undefined;
+
+    public layersNeedUpdate: boolean;
+
+    public override defines: LayeredMaterialDefines;
+
+    constructor(options: LayeredMaterialParameters = {}, crsCount: number) {
+        super(options);
```

Comment:
If we wish to let the user provides its own uniforms, it would be better to assign them directly in this class.
```suggestion
        const { uniforms, ...materialOptions } = options;
        super(materialOptions);
```

This would remove the need (in other part of the code) to check if the uniform is already assigned.